libdnf/rpm/package_query.cpp

```diff
@@ -2453,6 +2456,56 @@ void PackageQuery::filter_recent(const time_t timestamp) {
     }
 }
 
+void PackageQuery::filter_unneeded() {
+    auto & pool = get_rpm_pool(p_impl->base);
+
+    // get a list of all installed packages
+    auto * installed_repo = pool->installed;
+    if (installed_repo == nullptr) {
+        (*p_impl).clear();
+        return;
+    }
+
+    libdnf::solv::SolvMap installed_solv_map(pool.get_nsolvables());
+    for (int it = installed_repo->start; it < installed_repo->end; ++it) {
+        installed_solv_map.add(it);
+    }
+
+    // find user-installed packages and add them to the solver job
+    libdnf::rpm::PackageSet installed_pset{p_impl->base, installed_solv_map};
+    libdnf::solv::IdQueue userinstalled;
+    for (const auto & pkg : installed_pset) {
+        auto reason = pkg.get_reason();
+        // package that is not dep / weak dep is considered user-installed here
+        if (reason != libdnf::transaction::TransactionItemReason::WEAK_DEPENDENCY &&
+            reason != libdnf::transaction::TransactionItemReason::DEPENDENCY) {
```

Comment:
Minor: The similar test is also done by SWDB (swdb.cpp line 479). The question is whether it would be better to have a one point for the decision for both logic?

Reply:
The question is whether it should be in Swdb. It's not used anywhere (I guess it is a dnf4 heritage) and now, with system state, to resolve userinstalled packages  swdb is not needed.

Reply:
I'll create a new `filter_userinstalled()` filter in PackageQuery and delete the unused (and I think even broken) code from Swdb.